src/main/java/frc/robot/systems/BallHandlingFSM.java

```diff
@@ -100,20 +108,24 @@ public void update(TeleopInput input) {
 	private FSMState nextState(TeleopInput input) {
 		switch (currentState) {
 			case IDLE:
-				if (input != null && input.isShooterButtonPressed()) {
+				if (input != null && pushCommandTime == -1 && input.isShooterButtonPressed()) {
 					pushCommandTime = Timer.getFPGATimestamp();
 
 					return FSMState.FIRING;
+				} else if (pushCommandTime != -1
+					&& Timer.getFPGATimestamp() - pushCommandTime > PUSH_TIME_SECONDS) {
+					pushCommandTime = -1;
+
+					return FSMState.RETRACTING;
 				} else {
 					return FSMState.IDLE;
 				}
 
 			case FIRING:
-				if (Timer.getFPGATimestamp() - pushCommandTime > PUSH_TIME_SECONDS) {
-					return FSMState.IDLE;
-				} else {
-					return FSMState.FIRING;
-				}
+				return FSMState.IDLE;
```

Comment:
doesn't this mean you'll always just return to IDLE immediately from FIRING/RETRACTING and set both solenoids to false? or am I understanding this wrong. would love to see an updated FSM

Reply:
Yes this is correct. The reason we do this, is that in order to set the direction of the solenoid, we must pulse the solenoid in that direction for only a few milliseconds, then power should be turned off, as it is undesired for the solenoid to constantly be receiving unnecessary power.